File: fixtures/scale/typescript/large/src/ciphers/Dualmask.ts
// Symmetric xor-mask cipher (dualmask).
import { Bytes } from '../util/Bytes';

export class Dualmask {
  static mask(): number[] {
    return [146, 44, 84];
  }

  static encrypt(text: string, key: number): string {
    const mask = Dualmask.mask();
    const codes = Bytes.toCodes(text).map((v, i) => v ^ mask[i % 3] ^ (key % 256));
    return Bytes.fromCodes(codes);
  }

  static decrypt(text: string, key: number): string {
    return Dualmask.encrypt(text, key);
  }
}

File: fixtures/scale/lua/large/ciphers/dualmask.lua
local bytes = require("util.bytes")

local function dualmask_mask()
  return { 146, 44, 84 }
end

local function dualmask_encrypt(text, key)
  local mask = dualmask_mask()
  local codes = bytes.to_codes(text)
  for i, c in ipairs(codes) do
    codes[i] = c ~ mask[((i - 1) % 3) + 1] ~ (key % 256)
  end
  return bytes.from_codes(codes)
end

local function dualmask_decrypt(text, key)
  return dualmask_encrypt(text, key)
end

return { dualmask_mask = dualmask_mask, dualmask_encrypt = dualmask_encrypt, dualmask_decrypt = dualmask_decrypt }
